libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2111,57 +2111,59 @@ void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCa
 		return;
 	}
 
-	auto const functionPointerType = dynamic_cast<FunctionTypePointer>(type(*arguments.front()));
-
-	if (!functionPointerType)
+	FunctionType const* externalFunctionType = nullptr;
+	if (auto const functionPointerType = dynamic_cast<FunctionTypePointer>(type(*arguments.front())))
+	{
+		// functionPointerType->kind() != FunctionType::Kind::DelegateCall
+		externalFunctionType = functionPointerType->asExternallyCallableFunction(false);
+	}
+	else
 	{
 		m_errorReporter.typeError(
 			5511_error,
 			arguments.front()->location(),
 			"Expected first argument to be a function pointer, not \"" +
-			type(*arguments.front())->canonicalName() +
+			type(*arguments.front())->toString() +
 			"\"."
 		);
 		return;
 	}
 
 	if (
-		functionPointerType->kind() != FunctionType::Kind::External &&
```

Comment:
Please do this check on the original type. Otherwise I think we would introduce a bug.

Reply:
Ah ok, it currently is the same thing. Then please add an assertion above that the kind is the same for both type pointers.